src/back/spv/writer.rs

```diff
@@ -885,10 +885,36 @@ impl Writer {
         output: &mut Vec<Instruction>,
     ) -> Instruction {
         match statement {
-            crate::Statement::Return { value: _ } => match function.return_type {
-                Some(_) => unimplemented!(),
-                None => Instruction::new(Op::Return),
-            },
+            crate::Statement::Return { value } => {
+                match function.return_type {
+                    Some(expected_ty) => {
+                        let value = value.expect(
+                            "Return with no value from function supposed to return a value",
+                        );
+
+                        // Parse the expression
+                        let value_expression = &function.expressions[value];
+                        let (value_id, value_ty) =
+                            self.parse_expression(ir_module, function, value_expression, output);
+
+                        // Check that the actual returned type matches the expected one
+                        let expected_ty = &ir_module.types[expected_ty].inner;
+                        assert_eq!(expected_ty, value_ty, "Type of returned expression does not match declared function return type");
+
+                        // Construct the return value instruction
+                        let mut instruction = Instruction::new(Op::ReturnValue);
+                        instruction.set_result(value_id);
```

Comment:
I'm not sure if `set_result` is correct here or it should be `add_operand`

Reply:
It should be set to `add_operand` indeed. 

The way I determine is by looking at the spec. The spec will say something like: `Result<id>`, so then you will use `set_result` function. When the spec mentions: `<id> Result Type`, you will use `set_type`. When both are not mentioned, it will be `set_operand`.